File: js/object-asset.js
/** encode action datatype */
var ObjectAsset = Class([Asset], {
    type: 'object',
    name_changed: function(self, type, from, to){
        /*if (type == 'image' && self.info.image == from){
            self.set_attr('image', to);
        }
        if (type == 'object') {
            if (self.info.parent === from) {
                self.set_attr('parent',to);
            }
            for (var event in self.info.events) {
                if (event.indexOf('collide_') === 0 && event.slice('collide_'.length) == from) {
                    self.change_event(event, 'collide_' + to);
                    event = 'collide_' + to;
                }
                var dirty = false;
                for (var i=0;i<self.info.events[event].length;i++) {
                    for (var attr in self.info.events[event][i][1]) {
                        if (self.info.events[event][i][1][attr][0] == 'object'){
                            if (self.info.events[event][i][1][attr][1] == from){
                                self.info.events[event][i][1][attr][1] = to;
                                dirty = true;
                            }
                        }
                    }
                }
                if (dirty) {
                    self.save_actions(event, self.info.events[event]);
                }
            }
        }*/
    },
    add_event: function(self, type) {
        if (self.info.events[type]){
            return false;
        }
        self.parent.ajax.queue('objects/add_event',{'id':self.info.id, 'event':type});
//        self.parent.ajax.send_queued('objects/add_event',{'name':self.info.name,'event':type});
        self.info.events[type] = [];
        return true;
    },
    remove_event:function(self,type){
        if (!self.info.events[type]){
            return false;
        }
        self.parent.ajax.queue('objects/remove_event', {'id':self.info.id, 'event':type});
//        self.parent.ajax.send_queued('objects/remove_event',{'name':self.info.name,'event':type});
        delete self.info.events[type];
        return true;
    },
    change_event:function(self, event, name, func){
        self.parent.ajax.queue('objects/change_event',{'id':self.info.id,'event':event,'nevent':name},function(res){
            func && func();
        });
        self.info.events[name] = self.info.events[event];
        delete self.info.events[event];
    },
    duplicate_event:function(self, orig, dup, func){
        self.parent.ajax.queue('objects/duplicate_event',{'id':self.info.id,'event':orig,'nevent':dup},function(res){
            self.info.events[dup] = res.actions;
            func();
        });
    },
    save_actions: function(self, event, actions) {
        self.info.events[event] = actions;
        var json = jsonify(self.info.events[event]);
        self.parent.ajax.queue('objects/save_actions',{'id':self.info.id,'event':event,'actions':json});
    }
});

File: cgi/base/objectasset.py
'''objects.py -- does most of the heavly lifting
here, handles the 'object' asset type

>>> id = new()['id']
>>> id
1
>>> id = clone(id)['id']
>>> id
2
>>> add_event(id, 'keydown')
>>> remove_event(id, 'mouseup')
Traceback (most recent call last):
...
KeyError: 'mouseup'
>>> load(id)['events']
{u'keydown': []}
>>> remove_event(id, 'keydown')
>>> add_event(id, 'keydown')
>>> save_actions(id, 'keydown', ['something'])
>>> duplicate_event(id, 'keydown', 'mouseup')
[u'something']
>>> events = load(id)['events']
>>> events['keydown'] == events['mouseup']
True

'''
import asset

from cgitools import exit,die
import drupal

#good: Jan 4

type = 'objects'
def new():
    return asset.new(type,{'visible':False,
        'image':None,'events':{},
        'attributes':{},'parent':'BaseObject'})

def clone(id):
    return asset.clone(type,id)

def load(id):
    return asset.load(type,id)

def rename(id,new):
    return asset.rename(type,id,new)

def delete(id):
    return asset.delete(type,id)

def save_order(order):
    return asset.save_items(type,order)

def set_attr(id, attr, value):
    return asset.set_attr(type, id, attr, value)

## try return the actions list via json??

def add_event(id,event):
  events = asset._get_attr('objects',id,'events')
  events[event] = []
  asset._set_attr('objects',id,'events', events)

def remove_event(id,event):
  events = asset._get_attr('objects',id,'events')
  del events[event]
  asset._set_attr('objects',id,'events', events)

def duplicate_event(id,event,nevent):
  events = asset._get_attr('objects',id,'events')
  events[nevent] = events[event]
  asset._set_attr('objects', id, 'events', events)
  return events[nevent]

def change_event(id,event,nevent):
  events = asset._get_attr('objects',id,'events')
  events[nevent] = events[event]
  del events[event]
  asset._set_attr('objects', id, 'events', events)
  return events[nevent]

def save_actions(id,event,actions):
  events = asset._get_attr('objects',id,'events')
  events[event] = actions
  asset._set_attr('objects',id,'events', events)

